### PR TITLE
Remove seedSelector usage in dev-setup

### PR DESCRIPTION
**Pull request by xrstf**
**How to categorize this PR?**

/area quality
/kind technical-debt

**What this PR does / why we need it**:
This is the first step towards #4078. This PR makes it so that the local dev setup does not rely on the catch-all seedSelector anymore.

To achieve this, the `start-gardenlet` script was adjusted to auto-detect the seed to use for the gardenlet. If more than one seed is found, the user has to set the `SEED_NAME` environment variable to select their desired seed cluster. The seed spec is merged into the `20-componentconfig-gardenlet` before starting the gardenlet. Likewise, the seed's kubeconfig is fetched from the garden cluster and placed in a temporary file in `dev/` as well.

I added the dependency on yq to produce nice looking YAML, so when the user wants to edit their gardenlet configuration, they don't have to do it in JSON. However, yq recently went from 3.x to 4.x and 4.x is completely incompatible with 3.x -- I chose to build this for yq 4, as it's a new dependency, but good luck to everyone like me who has yq3 and lots and lots of scripts that need the old yq.

**Release note**:

```breaking developer
`make start-gardenlet` does not use seedSelector anymore, making the dev gardenlet single-seed only. If you have multiple Seeds in your local setup, you can specify the seed to act on via the `SEED_NAME` make variable (e.g. `make start-gardenlet SEED_NAME=local-foo`).
```

### CODE REVIEW

File: charts/gardener/gardenlet/charts/runtime/templates/configmap-componentconfig.yaml [hunk 1]
```diff
@@ -161,10 +161,6 @@ data:
     {{- if .Values.global.gardenlet.config.featureGates }}
     featureGates:
 {{ toYaml .Values.global.gardenlet.config.featureGates | indent 6 }}
-    {{- end }}
```
Comment (timebertt) [inline charts/gardener/gardenlet/charts/runtime/templates/configmap-componentconfig.yaml]: Let's not remove this in the "production charts" yet (should be in another PR, I guess)

File: charts/gardener/gardenlet/values.yaml [hunk 2]
```diff
@@ -155,7 +155,6 @@ global:
       #       labels:
       #         istio: ingressgateway-handler-2
 
-    # seedSelector: {}
```
Comment (timebertt) [inline charts/gardener/gardenlet/values.yaml]: Same here.

File: hack/install-requirements.sh [hunk 3]
```diff
@@ -31,6 +31,12 @@ if [[ ${platform} == "Linux" ]]; then
     curl -L -o /usr/local/bin/jq https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64
     chmod +x /usr/local/bin/jq
   fi
+
+  if ! which yq &>/dev/null; then
```
Comment (timebertt) [inline hack/install-requirements.sh]: Proposal:
Let's not install `yq` to `/usr/local/bin` but rather to a repo-local tools bin dir and use it for the make targets / hack scripts from there.
I find it quite annoying that `make install-requirements` overwrites your global tools (e.g. helm), especially if there are breaking changes between your global version and the version required by gardener, which is why I never use this target. It is also used in our pipelines, where it makes sense, but it's just annoying for developers.

We can use the same approach as controller-runtime, see [here](https://github.com/kubernetes-sigs/controller-runtime/blob/985e819db7af1cd17782ca7ad27644d0ca655054/Makefile#L38-L42) and respective [targets](https://github.com/kubernetes-sigs/controller-runtime/blob/985e819db7af1cd17782ca7ad27644d0ca655054/Makefile#L64-L77).

Had this on my list for a long time, but haven't been able to do it yet. But as you mentioned there are many breaking changes between yq 3 and 4, let's use this opportunity to at least install yq in a proper tools bin dir.
WDYT?

File: hack/local-development/start-gardenlet [hunk 4]
```diff
@@ -29,19 +31,69 @@ else
...
+      names="$(echo "$seeds" | yq eval --tojson --indent 0 '[.[].metadata.name]' -)"
+      echo "Multiple seeds exist in the garden, please specify \$SEED_NAME to select one of $names"
+      exit 1
+      ;;
+  esac
+fi
+
+# cleanup a bit to make the generated configuration file look a bit neater
+tmpFile="$DEV_DIR/seed-$SEED_NAME.yaml"
+echo "$seedConfig" | yq eval 'del(.apiVersion,.kind,.status,.metadata.annotations,.metadata.managedFields)' - > "$tmpFile"
```
Comment (timebertt) [inline hack/local-development/start-gardenlet]: Can you also cleanup `metadata.{creationTimestamp,finalizers,generation,resourceVersion,uid}`?

File: hack/local-development/start-gardenlet [hunk 5]
```diff
@@ -29,19 +31,69 @@ else
...
+configFile="$DEV_DIR/20-componentconfig-gardenlet-seed-$SEED_NAME.yaml"
+
+if [ ! -f "$configFile" ]; then
+  cp "$DEV_DIR/20-componentconfig-gardenlet.yaml" "$configFile"
+fi
+
+# update gardenlet config
+yq eval-all --inplace 'select(fileIndex==0).seedConfig = select(fileIndex==1) | select(fileIndex==0)' "$configFile" "$tmpFile"
+
+# fetch seed kubeconfig
```
Comment (timebertt) [inline hack/local-development/start-gardenlet:85]: We probably need some error handling (warning, explanation) in case the developer has not given `.spec.secretRef`.

**Review by timebertt — CHANGES_REQUESTED** [review]
Very nice, I like it!
A bit of feedback inline.

File: hack/install-requirements.sh [hunk 6]
```diff
@@ -24,6 +24,9 @@ export GO111MODULE=on
 curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.27.0
 curl -s "https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3" | bash -s -- --version 'v3.5.4'
 
+localBin="$(dirname "$0")/tools/bin"
```
Comment (timebertt) [inline hack/install-requirements.sh]: Why did you choose to install it here in bash instead of directly in the makefile (see [reference](https://github.com/kubernetes-sigs/controller-runtime/blob/985e819db7af1cd17782ca7ad27644d0ca655054/Makefile#L64-L77)?
I think the makefile approach is quite elegant and also ergonomic, as you don't have to worry about running `make install-requirements` or anything else first. It just ensures, that the needed tools are present when and you can directly run `make start-gardenlet`.

File: hack/install-requirements.sh [hunk 7]
```diff
@@ -32,10 +35,10 @@ if [[ ${platform} == "Linux" ]]; then
     chmod +x /usr/local/bin/jq
   fi
 
-  if ! which yq &>/dev/null; then
+  if ! [ -x "$localBin/yq" ]; then
     echo "Installing yq ..."
-    curl -L -o /usr/local/bin/yq https://github.com/mikefarah/yq/releases/download/v4.9.6/yq_linux_amd64
-    chmod +x /usr/local/bin/yq
+    curl -L -o "$localBin/yq" https://github.com/mikefarah/yq/releases/download/v4.9.6/yq_linux_amd64
```
Comment (timebertt) [inline hack/install-requirements.sh]: Oh BTW, many of us run on macOS (and we also had a few win based contributors). So we should support all 3 platforms with the installation.

File: hack/local-development/start-gardenlet [hunk 8]
```diff
@@ -29,19 +31,90 @@ else
...
+  .metadata.generation,
+  .metadata.resourceVersion,
+  .metadata.uid
+)' - > "$tmpFile"
+
+# create seed-specific configuration file if it doesn't exist yet
+configFile="$DEV_DIR/20-componentconfig-gardenlet-seed-$SEED_NAME.yaml"
+
+if [ ! -f "$configFile" ]; then
+  cp "$DEV_DIR/20-componentconfig-gardenlet.yaml" "$configFile"
```
Comment (timebertt) [inline hack/local-development/start-gardenlet]: Let's copy from the examples dir. Otherwise devs that have a `seedSelector={}` in their `dev/20-com...` (which is currently the default), will be unable to start the gardenlet and it fails with a long message including `either seed config or seed selector is required`.
```suggestion
  cp "$EXAMPLE_DIR/20-componentconfig-gardenlet.yaml" "$configFile"
```


Other option: keep the copy but make sure to remove the `seedSelector` from it.

File: hack/local-development/start-gardenlet [hunk 9]
```diff
@@ -29,19 +31,90 @@ else
   trap cleanup_kubeconfig EXIT
 fi
 
+YQ="$REPO_ROOT/hack/tools/bin/yq"
+GARDEN_KUBECONFIG="${GARDEN_KUBECONFIG:-$kubeconfig}"
+
+# determine the seed to manage; if multiple seeds exist on the
+# garden, the SEED_NAME env variable must be set
+if [ -n "${SEED_NAME:-}" ]; then
```
Comment (timebertt) [inline hack/local-development/start-gardenlet:39]: Proposal: 
Add a `SEED_NAME` var in `Makefile` that passes down to the env var if set, so that one can also run `make start-gardenlet SEED_NAME=my-seed`. This is a bit more obvious from the makefile already, shell completion can kick in if you've set it up, ...

File: hack/local-development/start-gardenlet [hunk 10]
```diff
@@ -17,7 +17,9 @@
 source $(dirname "${0}")/common/helpers
 source $(dirname "${0}")/common/local-imagevector-overwrite
 
-REPO_ROOT=$(dirname $0)/../..
+REPO_ROOT=$(dirname "${0}")/../..
```
Comment (timebertt) [inline hack/local-development/start-gardenlet:20]: Please adapt https://github.com/gardener/gardener/blob/master/docs/development/local_setup.md#run-the-gardener accordingly.

**Review by timebertt — COMMENTED** [review]
Thanks for updating the PR. Tested it and played around with it a bit.
Few more comments inline, otherwise looks quite good.

File: hack/install-requirements.sh (hunk 7 above)
Comment (xrstf) [reply hack/install-requirements.sh]: So the entire installation for Linux, Windows and Mac must be done inside the Makefile?

Comment (xrstf) [reply hack/install-requirements.sh]: Or should the `start-gardenlet` task simply require the `install-requirements` task? If not, should it require the `YQ` task? And if so, can that be a script? Would it need to be Bash+Powershell to support Windows?

File: hack/local-development/start-gardenlet (hunk 8 above)
Comment (timebertt) [reply hack/local-development/start-gardenlet]: Hmm, I think, this is not resolved.

File: Makefile [hunk 11]
```diff
@@ -27,11 +27,29 @@ REPO_ROOT                              := $(shell dirname $(realpath $(lastword
...
 ifneq ($(strip $(shell git status --porcelain 2>/dev/null)),)
 	EFFECTIVE_VERSION := $(EFFECTIVE_VERSION)-dirty
 endif
 
+#########################################
+# Binaries                              #
+#########################################
+
+$(YQ):
+	curl -L -o "$(YQ)" https://github.com/mikefarah/yq/releases/download/v4.9.6/yq_$(OS)_$(ARCH)
```
Comment (timebertt) [inline Makefile:51]: Either add `--create-dirs` or a `mkdir -p $(TOOLS_BIN_DIR)` somewhere, otherwise it will fail, if devs haven't created the dir beforehand.

**Review by timebertt — COMMENTED** [review]
One more request :)

**Comment by xrstf** [comment]
Done :)

**Review by timebertt — APPROVED** [review]
Nice, thank you very much!
Works like a charm 🎉 
/lgtm
/merge squash
